### PR TITLE
[server, db] Performance improvements to OAuthProvider sync

### PR DESCRIPTION
**Pull request by geropl**
## Description
Three optimizations:
 - all requests to `/auth/...` need to check which host provider is the right one, by matching a string. This parsed a `URL` on the fly each time for each entry, making it rather expensive
 - we regularly (each 2s) poll the with `getAllAuthProviders` to sync with the in-memory state. This is expensive, because we need to materialize the (encrypted) oauth config, despite not using it 99.99% of cases. This change introduces a field `oauthRevision` besides the `oauth` field that is used to filter already known objects. As a result, the list of auth providers to sync should stay empty in most cases.
 - make sure to `dispose()` `setTimout` on websocket ping-pong in every case
 
@AlexTugarev Would love to discuss this before Christmas. No need to finish/merge this PR before, though.

## Related Issue(s)

Context: #7082

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#7082*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-db/src/typeorm/auth-provider-entry-db-impl.ts
Comment (AlexTugarev) [inline components/gitpod-db/src/typeorm/auth-provider-entry-db-impl.ts:53]: this is new since last typeorm bump. 👏🏻  need to rewrite a handful of workarounds we've used before that. 

please check, that this actually works, according to the docs it's missing `...`, see https://github.com/typeorm/typeorm/blob/master/docs/select-query-builder.md#using-parameters-to-escape-data

File: components/server/src/auth/host-context-provider-impl.ts
Comment (AlexTugarev) [inline components/server/src/auth/host-context-provider-impl.ts:77]: nit: it's not just new entries, but also updated ones.

**Comment by AlexTugarev** [comment]
> make sure to dispose() setTimout on websocket ping-pong in every case

that's not included here, right?

@geropl, there is only one question about the query builder and the `WHERE ... IN ..` syntax. ok, that's also crucial for the intended optimization. once verified, we should merge, the changes look great! thanks for the effort digging through it.

**Comment by geropl** [comment]
@AlexTugarev Done with fixes and tests. I did not test the integration, yet. Will do on Monday with [in-workspace GitLab](https://gitpod.slack.com/archives/C02EN94AEPL/p1643926436870879).

**Comment by geropl** [comment]
@AlexTugarev done :point_up:

**Review by AlexTugarev — APPROVED** [review]

**Comment by geropl** [comment]
/unhold

Uah, missed this handbrake... :grimacing:

**Pull request merged**
